src/pool/srv_pool.c

```diff
@@ -4266,6 +4414,180 @@ ds_pool_update(uuid_t pool_uuid, crt_opcode_t opc,
+	rc = pool_map_extend(map, map_version, map_buf);
+	if (rc != 0)
+		D_GOTO(out_map, rc);
+
+	/* Write the new pool map. */
+	rc = pool_buf_extract(map, &map_buf);
+	if (rc != 0)
+		D_GOTO(out_map, rc);
+
+	rc = write_map_buf(tx, &svc->ps_root, map_buf, map_version);
+	if (rc != 0)
+		D_GOTO(out_map, rc);
+
+	rc = rdb_tx_commit(tx);
+	if (rc != 0) {
+		D_DEBUG(DB_MD, DF_UUID": failed to commit: "DF_RC"\n",
+			DP_UUID(svc->ps_uuid), DP_RC(rc));
+			D_GOTO(out_map, rc);
+	}
+
+	updated = true;
+
+	/* Update svc->ps_pool to match the new pool map. */
+	rc = ds_pool_tgt_map_update(svc->ps_pool, map_buf, map_version);
+	if (rc != 0) {
+		/*
+		* We must resign to avoid handling future requests with a
+		* stale pool map cache.
+		*/
+		rdb_resign(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term);
+		rc = 0;
+		goto out_map;
+	}
+
+	ds_rsvc_request_map_dist(&svc->ps_rsvc);
+
+out_map:
+	if (map_version_p != NULL)
+		*map_version_p = pool_map_get_version((map == NULL || rc != 0) ?
+						      svc->ps_pool->sp_map :
+						      map);
+	ABT_rwlock_unlock(svc->ps_lock);
+	rdb_tx_end(tx);
+
+out_uuids:
+	D_FREE(uuids);
+
+out_map_buf:
+	pool_buf_free(map_buf);
+
+	if (updated_p)
+		*updated_p = updated;
+	if (map)
+		pool_map_decref(map);
+
+	return rc;
+}
+
+int
+ds_pool_extend(uuid_t pool_uuid, struct rsvc_hint *hint, uint32_t nnodes,
+		uuid_t target_uuids[], d_rank_list_t *rank_list,
+		uint32_t ndomains, int32_t *domains, uint32_t *map_version_p)
+
+{
+	struct pool_svc		*svc;
+	struct rdb_tx		tx;
+	bool			updated;
+	int rc;
+
+	rc = pool_svc_lookup_leader(pool_uuid, &svc, hint);
+	if (rc != 0)
+		D_GOTO(out_svc, rc);
+
+	rc = rdb_tx_begin(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term, &tx);
+	if (rc != 0)
+		D_GOTO(out_svc, rc);
+	ABT_rwlock_wrlock(svc->ps_lock);
+
+	rc = ds_pool_extend_internal(&tx, svc, ndomains, target_uuids,
+		rank_list, ndomains, domains,
+		&updated, map_version_p, hint);
+
+	if (!updated)
+		D_GOTO(out_svc, rc);
+
+	/* Do rebuild stuff here */
+	/* Do rebuild stuff here */
+	/* Do rebuild stuff here */
+	/* Do rebuild stuff here */
+
+out_svc:
+	if (hint != NULL)
+		ds_rsvc_set_hint(&svc->ps_rsvc, hint);
+	pool_svc_put_leader(svc);
+	return rc;
+}
+
+void
+ds_pool_extend_handler(crt_rpc_t *rpc)
+{
+	struct pool_extend_in	*in = crt_req_get(rpc);
+	struct pool_extend_out	*out = crt_reply_get(rpc);
+	uuid_t			pool_uuid;
+	uuid_t			*target_uuids;
+	d_rank_list_t		rank_list;
+	uint32_t		ndomains;
+	int32_t			*domains;
+	int rc = 0;
+
+	uuid_copy(pool_uuid, in->pei_op.pi_uuid);
+	target_uuids = in->pei_tgt_uuids.ca_arrays;
+	rank_list.rl_nr = in->pei_tgt_ranks->rl_nr;
+	rank_list.rl_ranks = in->pei_tgt_ranks->rl_ranks;
+	ndomains = in->pei_ndomains;
+	domains = in->pei_domains.ca_arrays;
+
+	rc = ds_pool_extend(pool_uuid, &out->peo_op.po_hint, ndomains,
+		       target_uuids, &rank_list, ndomains, domains,
+		       &out->peo_op.po_map_version);
+
+	if (target_uuids == NULL || ndomains || domains || rank_list.rl_nr > 0)
+		goto out;
```

Comment:
hmm, what does this if check mean?

Reply:
Sorry, dead code that I missed. removed in most recent patch.